File: leet-code/937. Reorder Data in Log Files/ReorderDataInLogFiles937_2.java
import java.util.ArrayList;
import java.util.List;

public class ReorderDataInLogFiles937_2 {
    public String[] reorderLogFiles(String[] logs) {
        List<String> digits = new ArrayList<>();
        List<String> letters = new ArrayList<>();

        for (String log : logs) {
            if (isDigitOnly(log)) {
                digits.add(log);
            } else {
                letters.add(log);
            }
        }

        letters.sort((first, second) -> {
            String[] firstWords = first.split(" ", 2);
            String[] secondWords = second.split(" ", 2);

            if (firstWords[1].compareTo(secondWords[1]) == 0) {
                return firstWords[0].compareTo(secondWords[0]);
            }
            return firstWords[1].compareTo(secondWords[1]);
        });

        letters.addAll(digits);
        return letters.toArray(new String[0]);
    }

    private boolean isDigitOnly(String log) {
        String[] words = log.split(" ", 2);
        return words[1]
                .replaceAll("^[\\s0-9]+", "")
                .isEmpty();
    }
}

File: leet-code/937. Reorder Data in Log Files/ReorderDataInLogFiles937_2.kt
package com.example

class ReorderDataInLogFiles937_2 {
    fun reorderLogFiles(logs: Array<String>): Array<String> {
        val letterLogs = mutableListOf<String>()
        val digitLogs = mutableListOf<String>()
        for (log in logs) {
            if (log.split(" ")[1][0].isDigit()) {
                digitLogs.add(log)
            } else {
                letterLogs.add(log)
            }
        }
        letterLogs.sortWith { o1, o2 ->
            val o1Sliced = o1.split(" ", limit = 2)
            val o2Sliced = o2.split(" ", limit = 2)
            if (o1Sliced[1] == o2Sliced[1]) {
                o1Sliced[0].compareTo(o2Sliced[0])
            } else {
                o1Sliced[1].compareTo(o2Sliced[1])
            }
        }
        letterLogs.addAll(digitLogs)
        return letterLogs.toTypedArray()
    }
}
